tensorflow_io/core/kernels/avro/atds/shuffle_handler.h

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifndef TENSORFLOW_IO_CORE_KERNELS_AVRO_ATDS_SHUFFLE_HANDLER_H_
+#define TENSORFLOW_IO_CORE_KERNELS_AVRO_ATDS_SHUFFLE_HANDLER_H_
+
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/types.pb.h"
+#include "tensorflow/core/lib/random/philox_random.h"
+#include "tensorflow/core/lib/random/random.h"
+#include "tensorflow/core/lib/random/random_distributions.h"
+#include "tensorflow_io/core/kernels/avro/atds/avro_block_reader.h"
+
+namespace tensorflow {
+namespace data {
+
+class ShuffleHandler {
+ public:
+  ShuffleHandler(mutex* mu) {
+    mu_ = mu;
+    ResetRngs();
+  }
+  void SampleBlocks(size_t batch_size, bool shuffle,
```

Comment:
I went through the shuffle algorithm which was posted in the FDSDataset Google doc. The purpose of sampling records randomly from different blocks is mentioned as
```
 In this way, the data from the latter blocks can be collected 
earlier.
```
I am assuming, you mean that the shuffle op can be completed within the prefetch phase itself? (please correct me if I misunderstood) What was the performance gap when using the standard `tf.data` prefetching and shuffle ops and disabling this custom shuffling?

Reply:
No, it still shuffles in the parsing phase. The problem with Avro is that Avro block does not have metadata to track the offsets of each Avro record in the block. It makes it impossible to jump to a random offset and decode a record at that offset. In other words, we can only read Avro block sequentially. This limitation adds complexity to shuffle. If we want to shuffle Avro records in an Avro block, we have to read all records sequentially and shuffle the intermediate results. It will introduce extra copy which hurts performance. 

In the AvroTensorDataset, we propose a shuffle algorithm that samples the number of records to read from each Avro block and merges the read records from multiple blocks to the outputs. In this way, we can still read Avro records sequentially without extra copy. The Avro block will be kept in memory until they are fully read. Hope it add the context about the shuffle algorithm.

Reply:
Thanks @plliao